### DISCUSSION TITLE
How to disable the watch before resetting the fields?

### DISCUSSION THREAD

**Discussion by khanakia** [react-hook-form/react-hook-form#10443] [Q&A]
Hi,

I do not want to run the `watch` useEffect hook on reset. I just want to reset the values without watching trigggered.

I have already tried `keepDefaultValues, keepDirty` none of the options are working. 

no matter what watch always runs

**Comment by joshacheson** [comment]
Heya. I have a similar (or the same) problem and am going to provide an example to be illustrative. It might also make sense to widen the topic here slightly - it occurs to me that when using at least `reset` and `setValue` that you may wish to not have "watch subscriptions" fire. The current APIs that appear to make this possible feel unintuitive, in particular that "reset" and other programmatic means of changing the form's value don't come with an explicit `type` in the `watch` callback. Rather, we read the _absence_ of a value for `type` (`type === undefined`). I think it would be preferable if we could be explicit about events such as `reset` or `setValue` that cause a watch subscription to fire, including those which are not idiomatic to all forms on the web. Alternatively, if we could opt to not fire `watch` callbacks when changing values programmatically, i.e. `reset({...}, {triggerWatch: false});`, that could work as well.

Having said all of this I'm going to lay out the behaviour I'm seeking, my best attempt at implementing it, and am more or less asking whether from the maintainer's point of view this is the best way to implement this sort of thing.

Here's a description of the behaviour I'm seeking:

Given a form with the following shape: `{field1: string, field2: string, field3: string}`

- When `field1` is given value `'value1'`, fill in `field2` with `'value2'`.
- When `field2` is given `'value2'`, fill in `field3` with `'value3'` (do this whether the user manually changes `field2` triggering `type === 'change'` or whether `field2` is filled in with the behaviour from the previous bullet.
- Do not do any of this "prefill" behaviour described in the previous 2 bullets when calling `reset({...})`.

I'm also implementing my current solution given a particular prior, which is that `useWatch` and `watch` are generally optimized for running during the render cycle, and should be used when intending to _display_ the value you are watching (this is noted in the docs and in video tutorials). However, `watch`'s "callback api" (`watch((data, {type, name}) => console.log(data, type, name))`) is appropriate for use cases where we're driving side effects.

Given all of this, here's a stackblitz implementing what i'm looking for: https://stackblitz.com/edit/stackblitz-starters-1xdi9r?file=src%2FApp.tsx. @khanakia I would want to note that following the patterns in this example, I think you may be able to get the behaviour you're looking for. They key part is checking for the `type` provided in the second arguments to our "watch callback". If the type is not "change", then don't do whatever behaviours it is you're trying to avoid in the case of `reset({...})`.

Anyways, given my example, here are the issues that I see and my commentary otherwise:

- It's not possible with the current API to do "cascading pre-fill" (changes to `field1` can pre-fill `field2`, and changes to `field2` will then pre-fill `field3`). This is because I'm guarding against doing this behaviour during `reset` by checking for `type === 'change'`. Note that I have to call my pre-fill function for `field3` in two blocks, the block which handles changes to `field1` _and_ the block which handles changes to `field2` (I would prefer to have this naturally cascade).
- The API for my pre-fill functions (`fillField2` in particular) has to return the new value for `field2`. Without this, my call to pre-fill `field3`'s value (on line 64) will not work, because the value from `data` inside the `watch` callback is not reflecting the updated value after my call to `setValue` inside `fillField2`.
- I feel unconfident about guarding against "pre-fill on reset" the way I'm doing it (by checking for `type === 'change'`

Anyways, this is my sense of the issues. I think I'm technically able to implement the behaviour I'm looking for, but struggling with whether this would be the recommended approach, and also wondering if it's possible to enhance existing APIs to make behaviour like this more natural or intuitive.

For what it's worth, I am a _huge_ fan of this library, I have enormous appreciation for the maintainer(s), and I'm not meaning to appear critical. Rather, I'm hoping to contribute to useful discussion.
